Abstraction and on disk implementation of blob storage

shubham3121 opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

shubham3121 committed:
brainstrom abstractions for blob storage

rasswanth-s commented:
Blob Storage 🙌 🙌 🙌  !!!

kiendang committed:
Add on disk impl

kiendang committed:
Some cleanup

kiendang committed:
Typo

kiendang committed:
Minor fix

kiendang committed:
Prelim on disk file store impl

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Lint

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Use pydantic to simplify the type

kiendang committed:
Fix api call

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Lint

kiendang committed:
Use pydantic for FileClientConfig as well

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Init node file client

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Remove privateattr

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Write object file abstraction

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Fix instance method

kiendang committed:
Fix up

kiendang committed:
Add file save service endpoint

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Reorganize to avoid circular imports

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Fix serializable decorator

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Add implementation for on disk storage

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Fix context manager

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Add example notebook

Co-authored-by: Shubham Gupta <shubhamgupta3121@gmail.com>

kiendang committed:
Attempt to fix notebook

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Attempt to fix OnDiskFileClient not in TYPE_BANK

kiendang committed:
Attempt to fix OnDiskFileClientConfig not in TYPE_BANK

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Make notebook a bit clearer

kiendang committed:
Tweak notebook

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Attempt to fix pathlib.PosixPath not in TYPE_BANK

kiendang committed:
Fix for pathlib.Path subtypes

kiendang committed:
Use more precise typing

kiendang committed:
Extend Path serde to include PurePath and its subtypes

kiendang committed:
Break down file_store.py into separate files

for interfaces and implementations

kiendang committed:
Convert comments into a docstring

kiendang committed:
Edit docstring

kiendang committed:
Add more doc

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Correct use of classmethod

kiendang committed:
Merge branch 'dev' into blob-storage

kiendang committed:
Rename file_object/file service/stash to blob_storage

kiendang committed:
Rename file_store to blob_storage

kiendang committed:
Rename SyftResource to BlobRetrieval

kiendang committed:
Rename SyftWriteResource to BlobDeposit

kiendang committed:
Should inherit from SyftResource instead of SyftObject

kiendang committed:
Rename types/file_object.py to types/blob_storage.py

kiendang committed:
Rename FileObject to BlobStorageEntry

kiendang committed:
Fix doc

kiendang commented:
@shubham3121 @PeterChung241 @khoaguin I've finished renaming all the file stuff to blob storage as we discussed. Please review all the commits starting with `Rename ...`.

There are 2 changes I made that are different from what we discussed. One is renaming `SyftResource`  and `SyftWriteResource` to `BlobRetrieval` and `BlobDeposit` instead of `SyftReadResource`  and `SyftWriteResource`. Alr discussed with Shubham about this.

The second one is renaming `FileObject` to `BlobStorageEntry` instead of `BlobStorageInformation`. Reasons are:
- We have things like `CreateFileObject`, `get_file_object_by_uid`, `get_all_file_object`, ... Replacing `file_object` here with `blob_storage_entry` instead of `blob_storage_information` reads better grammatically
- The object works like an entry in the blob storage as well since it serves as a record to help allocate and keep track of the syft object in the blob storage.

Let me know what you think. Happy to make changes again.

kiendang committed:
NIT

kiendang commented:
Sorry for abusing force pushing, just to make the commits clean for easier review. If you want to synchronize your local repo with this PR do `git pull --rebase`.

kiendang committed:
Shorten BlobStorageConfig attribute names

to be less redundant and also more in line with other configs (e.g. QueueConfig)

kiendang committed:
Correct typing

shubham3121 committed:
Merge branch 'dev' into blob-storage

shubham3121 committed:
handle error on blob storage entry set

shubham3121 commented on packages/syft/src/syft/store/blob_storage/__init__.py:
```diff
@@ -0,0 +1,135 @@
+"""Blob file storage
+
+Contains blob file storage interfaces. See `on_disk.py` for an example of a concrete implementation.
```
❤️

shubham3121 commented:
Excellent work @kiendang !! 💯 🔥 ❤️

kiendang reviewed (approved):
@shubham3121 I realized you created this PR so probably can't approve it? I'm gonna just approve it here. When everything looks good you can merge it.
